ext/console/02_console.js

```diff
@@ -390,61 +390,85 @@
     const entries = [];
     let iter;
     let valueIsTypedArray = false;
+    let entriesLength;
 
     switch (options.typeName) {
       case "Map":
         iter = MapPrototypeEntries(value);
+        entriesLength = value.size;
         break;
       case "Set":
         iter = SetPrototypeEntries(value);
+        entriesLength = value.size;
         break;
       case "Array":
-        iter = ArrayPrototypeEntries(value);
+        entriesLength = value.length;
         break;
       default:
         if (isTypedArray(value)) {
+          entriesLength = value.length;
           iter = ArrayPrototypeEntries(value);
           valueIsTypedArray = true;
         } else {
           throw new TypeError("unreachable");
         }
     }
 
-    let entriesLength = 0;
-    const next = () => {
-      return iter.next();
-    };
-    while (true) {
-      let el;
-      try {
-        const res = iter.next();
-        if (res.done) {
-          break;
-        }
-        el = res.value;
-      } catch (err) {
-        if (valueIsTypedArray) {
-          // TypedArray.prototype.entries doesn't throw, unless the ArrayBuffer
-          // is detached. We don't want to show the exception in that case, so
-          // we catch it here and pretend the ArrayBuffer has no entries (like
-          // Chrome DevTools does).
-          break;
-        }
-        throw err;
-      }
-      if (entriesLength < inspectOptions.iterableLimit) {
+    if (options.typeName === "Array") {
+      for (
+        let i = 0, j = 0;
+        i < value.length && j < inspectOptions.iterableLimit;
+        i++, j++
+      ) {
         inspectOptions.indentLevel++;
-        ArrayPrototypePush(
-          entries,
-          options.entryHandler(
-            el,
-            inspectOptions,
-            FunctionPrototypeBind(next, iter),
-          ),
+        const { entry, skipTo } = options.entryHandler(
+          [i, value[i]],
+          inspectOptions,
         );
+        ArrayPrototypePush(entries, entry);
         inspectOptions.indentLevel--;
+
+        if (skipTo) {
+          // substract skipped (empty) items
+          entriesLength -= skipTo - i;
+          i = skipTo;
+        }
+      }
+    } else {
+      let i = 0;
+      while (true) {
+        let el;
+        try {
+          const res = iter.next();
+          if (res.done) {
+            break;
+          }
+          el = res.value;
+        } catch (err) {
+          if (valueIsTypedArray) {
+            // TypedArray.prototype.entries doesn't throw, unless the ArrayBuffer
+            // is detached. We don't want to show the exception in that case, so
```

Comment:
Although this was already here, this should only throw once, before the first iteration - `typed_array.entries()` is what throws, but the `try` is around the `.next()`?

(It's only possible that it would throw if `byteLength = length = 0`, so it ought to special-case the empty TypedArray object before calling it.)

But looking around more, on line 410, `Array#entries` is used instead of `TypedArray#entries` anyways?

Reply:
Dw, there's no issue with your work.

I'll open an issue about it when this PR is landed. The optimisation for comparing the length against zero should be applicable here afterwards.